File: plugins/agent-capability-kit/skills/plugin-authoring/assets/servers/example-mcp.py
#!/usr/bin/env -S uv run
# /// script
# dependencies = []
# ///

"""Example Model Context Protocol (MCP) server implementation."""

from typing import Any

import json
import os
import sys


def read_message() -> dict[str, Any] | None:
    """Read a message from stdin with MCP headers."""
    headers = {}
    while True:
        line = sys.stdin.buffer.readline()
        if not line:
            return None
        if line in (b"\r\n", b"\n"):
            break
        name, value = line.decode("utf-8").split(":", 1)
        headers[name.lower()] = value.strip()
    length = int(headers.get("content-length", "0"))
    if length == 0:
        return None
    body = sys.stdin.buffer.read(length)
    return json.loads(body.decode("utf-8"))


def write_message(message: dict[str, Any]) -> None:
    """Write a message to stdout with MCP headers."""
    encoded = json.dumps(message).encode("utf-8")
    sys.stdout.buffer.write(
        f"Content-Length: {len(encoded)}\r\n\r\n".encode(),
    )
    sys.stdout.buffer.write(encoded)
    sys.stdout.buffer.flush()


def success(message_id: Any, result: Any) -> dict[str, Any]:
    """Create a successful JSON-RPC response."""
    return {"jsonrpc": "2.0", "id": message_id, "result": result}


def error(
    message_id: Any,
    code: int,
    message: str,
) -> dict[str, Any]:
    """Create a JSON-RPC error response."""
    return {
        "jsonrpc": "2.0",
        "id": message_id,
        "error": {"code": code, "message": message},
    }


def tool_definition() -> dict[str, Any]:
    """Return the tool definition for read_plugin_paths."""
    return {
        "name": "read_plugin_paths",
        "description": (
            "Return the plugin root and data directories for local runtime checks."
        ),
        "inputSchema": {
            "type": "object",
            "properties": {},
            "additionalProperties": False,
        },
    }


def tool_result() -> dict[str, Any]:
    """Return the result of the read_plugin_paths tool."""
    plugin_root = os.environ.get("CLAUDE_PLUGIN_ROOT", "")
    plugin_data = os.environ.get("CLAUDE_PLUGIN_DATA", "")
    return {
        "content": [
            {
                "type": "text",
                "text": json.dumps(
                    {
                        "pluginRoot": plugin_root,
                        "pluginData": plugin_data,
                    },
                    indent=2,
                ),
            },
        ],
    }


def handle_request(message: dict[str, Any]) -> dict[str, Any] | None:
    """Handle an MCP request message."""
    method = message.get("method")
    message_id = message.get("id")
    if method == "initialize":
        return success(
            message_id,
            {
                "protocolVersion": "2024-11-05",
                "serverInfo": {"name": "example-local-mcp", "version": "0.1.0"},
                "capabilities": {"tools": {}},
            },
        )
    if method == "notifications/initialized":
        return None
    if method == "tools/list":
        return success(message_id, {"tools": [tool_definition()]})
    if method == "tools/call":
        params = message.get("params", {})
        if params.get("name") != "read_plugin_paths":
            return error(message_id, -32602, "Unknown tool")
        return success(message_id, tool_result())
    if method == "shutdown":
        return success(message_id, {})
    return error(message_id, -32601, f"Unsupported method: {method}")


def main() -> None:
    """Run the MCP server."""
    while True:
        message = read_message()
        if message is None:
            break
        if message.get("method") == "exit":
            break
        response = handle_request(message)
        if response is not None:
            write_message(response)


if __name__ == "__main__":
    main()

File: plugins/agent-capability-kit/skills/plugin-authoring/assets/servers/example-mcp.ts
#!/usr/bin/env bun
// -*- coding: utf-8 -*-

import { createInterface } from "node:readline";

type JsonValue =
  | null
  | boolean
  | number
  | string
  | readonly JsonValue[]
  | { readonly [key: string]: JsonValue };

type JsonObject = Readonly<Record<string, JsonValue>>;

const PROTOCOL_VERSION = "2025-11-25";

/** Return true when a parsed value is a JSON object. */
const isJsonObject = (value: unknown): value is JsonObject =>
  typeof value === "object" && value !== null && !Array.isArray(value);

/** Read a string property from a JSON object. */
const stringField = (message: JsonObject, key: string): string => {
  const value = message[key];
  return typeof value === "string" ? value : "";
};

/** Write one newline-delimited JSON-RPC message to stdout. */
const writeMessage = (message: JsonObject): void => {
  process.stdout.write(`${JSON.stringify(message)}\n`);
};

/** Create a successful JSON-RPC response. */
const success = (
  messageId: JsonValue | undefined,
  result: JsonValue
): JsonObject => ({ id: messageId ?? null, jsonrpc: "2.0", result });

/** Create a JSON-RPC error response. */
const error = (
  messageId: JsonValue | undefined,
  code: number,
  message: string
): JsonObject => ({
  error: { code, message },
  id: messageId ?? null,
  jsonrpc: "2.0"
});

/** Return the tool definition for read_plugin_paths. */
const toolDefinition = (): JsonObject => ({
  description:
    "Return the plugin root and data directories for local runtime checks.",
  inputSchema: {
    additionalProperties: false,
    properties: {},
    type: "object"
  },
  name: "read_plugin_paths"
});

/** Return the result of the read_plugin_paths tool. */
const toolResult = (): JsonObject => ({
  content: [
    {
      text: JSON.stringify(
        {
          pluginData: process.env["CLAUDE_PLUGIN_DATA"] ?? "",
          pluginRoot: process.env["CLAUDE_PLUGIN_ROOT"] ?? ""
        },
        null,
        2
      ),
      type: "text"
    }
  ]
});

/** Handle one MCP request or notification. */
const handleMessage = (message: JsonObject): JsonObject | undefined => {
  const method = stringField(message, "method");
  const messageId = message["id"];
  const isNotification = messageId === undefined;
  if (method === "initialize") {
    return isNotification
      ? undefined
      : success(messageId, {
          capabilities: { tools: {} },
          protocolVersion: PROTOCOL_VERSION,
          serverInfo: { name: "example-mcp", version: "0.1.0" }
        });
  }
  if (method === "notifications/initialized") {
    return undefined;
  }
  if (method === "ping") {
    return isNotification ? undefined : success(messageId, {});
  }
  if (method === "tools/list") {
    return isNotification
      ? undefined
      : success(messageId, { tools: [toolDefinition()] });
  }
  if (method === "tools/call") {
    const { params } = message;
    if (isNotification) {
      return undefined;
    }
    if (
      !isJsonObject(params) ||
      stringField(params, "name") !== "read_plugin_paths"
    ) {
      return error(messageId, -32_602, "Unknown tool");
    }
    return success(messageId, toolResult());
  }
  return isNotification
    ? undefined
    : error(messageId, -32_601, `Unsupported method: ${method}`);
};

/** Read newline-delimited JSON-RPC messages from stdin. */
const runServer = async (): Promise<void> => {
  const lines = createInterface({
    crlfDelay: Number.POSITIVE_INFINITY,
    input: process.stdin
  });
  for await (const line of lines) {
    if (line.length === 0) {
      continue;
    }
    try {
      const parsed: unknown = JSON.parse(line);
      if (!isJsonObject(parsed)) {
        writeMessage(
          error(null, -32_600, "JSON-RPC message must be an object")
        );
        continue;
      }
      const response = handleMessage(parsed);
      if (response !== undefined) {
        writeMessage(response);
      }
    } catch {
      writeMessage(error(null, -32_700, "Invalid JSON"));
    }
  }
};

await runServer();
